pkg/agent/openflow/pipeline.go

```diff
@@ -1125,14 +1122,26 @@ func (c *client) traceflowL2ForwardOutputFlows(dataplaneTag uint8, liveTraffic,
 	return flows
 }
 
-// l2ForwardOutputServiceHairpinFlow uses in_port action for Service
-// hairpin packets to avoid packets from being dropped by OVS.
-func (c *client) l2ForwardOutputServiceHairpinFlow() binding.Flow {
-	return c.pipeline[L2ForwardingOutTable].BuildFlow(priorityHigh).
-		MatchRegMark(HairpinRegMark).
-		Action().OutputInPort().
-		Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
-		Done()
+// l2ForwardOutputServiceHairpinFlows uses in_port action for hairpin Service
+// packets to avoid packets from being dropped by OVS.
+// For request packets of hairpin Service connections, CT_MARK value set in
+// SNAT ct zone is used to match the packets. For reply packets, since CT_MARK
+// value set in SNAT ct zone will be lost after passing DNAT ct zone, before
+// passing DNAT ct zone, HairpinRegMark is set and is used to match the reply
+// packets.
+func (c *client) l2ForwardOutputServiceHairpinFlows() []binding.Flow {
```

Comment:
It's not only for the first packet of hairpin connections. It's also for the third, fifth .. packets (requests packets except the first one). The request packets (except the first one) cannot get the value of CT_MARK when passing through SNAT ct zone on table `snatConntrackTable` (but reply packets can get). When passing through SNAT ct zone on table `snatConntrackCommitTable`, the value of CT_MARK can be got. However, the next table `L2ForwardingOutTable`. I think we'd better add a flow rather than a new table.

Reply:
Actually I don't prefer to use ct mark in `L2ForwardingOutTable`, as I thought this table is possible to focus on L2 operations... Since the target ofport number is decided after L2FwdCalcTable, is it possible to set the regMark after that table?